extension/flat_tensor/test/test_serialize.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/extension/flat_tensor/serialize/flat_tensor_header.h>
+#include <executorch/extension/flat_tensor/serialize/scalar_type_generated.h>
+#include <executorch/extension/flat_tensor/serialize/schema_generated.h>
+#include <executorch/extension/flat_tensor/serialize/serialize.h>
+
+#include <gtest/gtest.h>
+
+#include <executorch/extension/tensor/tensor_ptr.h>
+#include <executorch/runtime/core/result.h>
+#include <executorch/runtime/platform/runtime.h>
+#include <map>
+
+#include <sstream>
+
+using namespace ::testing;
+using executorch::runtime::Error;
+using executorch::runtime::Result;
+
+class FlatTensorSerializeTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    // Since these tests cause ET_LOG to be called, the PAL must be initialized
+    // first.
+    executorch::runtime::runtime_init();
+  }
+};
+
+TEST_F(FlatTensorSerializeTest, ValidFlatTensorSerialized) {
+  const size_t kTensorAlignment = 16;
+  std::map<std::string, exec_aten::Tensor> flat_tensor_map;
+
+  float linear_weight = 3.14f;
+  auto weight = executorch::extension::make_tensor_ptr({1}, &linear_weight);
+
+  float linear_bias = 2.0f;
+  auto bias = executorch::extension::make_tensor_ptr({1}, &linear_bias);
+
+  flat_tensor_map.insert({"linear.weight", *weight.get()});
+  flat_tensor_map.insert({"linear.bias", *bias.get()});
+
+  std::ostringstream buf;
+  auto err = executorch::extension::flat_tensor::save_ptd(
+      buf, flat_tensor_map, kTensorAlignment);
+  ASSERT_EQ(err, Error::Ok);
+  auto x = buf.str();
+  const char* byte_buffer = x.c_str();
+
+  // Check Magic
+  ASSERT_EQ(byte_buffer[0], 'F');
+  ASSERT_EQ(byte_buffer[1], 'H');
+  ASSERT_EQ(byte_buffer[2], '0');
+  ASSERT_EQ(byte_buffer[3], '1');
+
+  // Check Header
+  ASSERT_EQ(
+      *(uint32_t*)(byte_buffer + 4),
+      executorch::extension::FlatTensorHeader::kHeaderExpectedLength);
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 8),
+      48); // Header is 40 bytes + 8 bytes of padding today, and then the
+           // flatbuffer starts.
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 16),
+      224); // This is fragile, and depends on the schema, the builder, and the
+            // padding needed.
+  const uint64_t segment_offset = 48 +
+      224; // Segment offset, depends on the padded header and flatbuffer sizes.
+  ASSERT_EQ(*(uint64_t*)(byte_buffer + 24), segment_offset);
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 32),
+      32); // Segment total size, 8 bytes of data (2 floats), 24 bytes of
```

Comment:
If there's only one segment, there shouldn't be any padding after it: there should only be padding between adjacent segments